keps/sig-storage/1472-storage-capacity-tracking/README.md

```diff
@@ -1129,33 +1154,31 @@ such a scenario, the scheduler has to make those decisions based on
 outdated information, in particular when making one scheduling
 decisions affects the next decision.
 
-We need to investigate:
-- Whether this really is a problem in practice, i.e. identify
-  workloads and drivers where this problem occurs.
-- Whether introducing some simple modeling of capacity helps.
-- Whether prioritization of nodes helps.
-
-For a discussion around modeling storage capacity, see the proposal to
-add ["total capacity" to
-CSI](https://github.com/container-storage-interface/spec/issues/301).
-
-### "Total available capacity" vs. "maximum volume size"
-
-The CSI spec around `GetCapacityResponse.capacity` [is
-vague](https://github.com/container-storage-interface/spec/issues/432)
-because it ignores fragmentation issues. The current Kubernetes API
-proposal follows the design principle that Kubernetes should deviate
-from the CSI spec as little as possible. It therefore directly copies
-that value and thus has the same issue.
-
-The proposed usage (comparison of volume size against available
-capacity) works either way, but having separate fields for "total
-available capacity" and "maximum volume size" would be more precise
-and enable additional features like even volume spreading by
-prioritizing nodes based on "total available capacity"
-
-The goal is to clarify that first in the CSI spec and then revise the
-Kubernetes API.
+[Scale testing](https://github.com/kubernetes-csi/csi-driver-host-path/blob/f053a7b0c4b719a5808fc47fdb3eba9cdade2067/docs/storage-capacity-tracking.md)
+showed that this can occur for a fake workload that generates
+pods with generic ephemeral inline volumes as quickly as possible: publishing
+CSIStorageCapacity objects was sometimes too slow, so scheduling retries were
```

Comment:
Can you somehow quantify this in the KEP?

> The ratio between "with volumes" and "no volumes" is 1.58.

Could we consider that the conclusion for the testing?

Although:

> VOL_SIZE: 1Gi
> --capacity=fast=100Gi

That doesn't seem like a particularly difficult scenario. kube-scheduler wouldn't already schedule more than 100 pods per nodes, plus kube-scheduler spreads by default. So I wouldn't expect that many retries anyways.

Can you try something like `--capacity=fast=20Gi`?

Reply:
> Could we consider that the conclusion for the testing?

What conclusion? That scheduling pods with volumes is 1.58 slower than for pods without volumes? That depends on so many other factors that I am not comfortable putting this into the KEP.

Also note that csi-driver-host-path as essentially no overhead internally for volumes. That is not realistic and (in this experiment) intentionally stresses the control plane and sidecar more than other drivers.

> That doesn't seem like a particularly difficult scenario. 

Scheduling 10000 pods (100 per node, 100 nodes) with 10000 volumes is not difficult? That sounds like a lot of volumes to me.

> Can you try something like --capacity=fast=20Gi?

And which volume size? If I keep it at 1Gi then I also need to reduce the number of pods, otherwise not all of them can be started. That will make the problem simpler, not harder.

I'm not sure what you are trying to achieve here.

Reply:
After thinking about this I think I understand what you mean: your theory is that the 100 pods per node limit influences the outcome and therefore you want to test with 20 pods per node. Makes sense. It doesn't matter whether we reduce the capacity per node or increase the volume size, I'll test with 5Gi volumes (simpler to configure).

Could it be that the Azure cloud is configured to allow more than 100 pods per node? If there really is such a limit, scheduling without storage capacity should have worked, but it didn't: retries were high for 10 nodes and the test timed out for 100 nodes.

Reply:
yes, I want to test the case where storage can allow significantly less pods than the actual number of pods limit.

I think the hard limit is 110 pods per node, unless they have a patched kubelet?

Reply:
So with 100 I was still below that limit. Then the results with the test timeout for 100 nodes without storage capacity tracking makes sense, because the scheduler may have tried to put more than 100 pods onto a single node.

I ran tests with 10 nodes and 20 pods. Now the setup without storage capacity tracking failed. Baseline was 29s, with volumes and capacity tracking 49s, a factor of 1.69 - similar to what was observed before. See https://github.com/kubernetes-csi/csi-driver-host-path/pull/351 for details.

Reply:
Thanks.

Numbers look acceptable. Maybe a bigger, more busy cluster could have more issues, but that's harder to proof.

The real problem will be when we need to scale up. I hope you can still bring this discussion back to CSI.

/lgtm